### PR TITLE
GUILess implementation

### PR DESCRIPTION
**Pull request by KilnOfTheSecondFlame**
Prepare for refactoring and tears 😅

### CODE REVIEW

File: src/main/kotlin/ApplicationStarter.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import gui.GuiLess
+import gui.ScannerAbstractionImpl
+import test.cases.dataloader.JsonDataLoader
+import util.ReadableFileImpl
+import java.io.File
+
+fun main() {
+    val scannerAbstractionImpl = ScannerAbstractionImpl()
+    val guiLess = GuiLess(scannerAbstractionImpl)
+    val jsonFile = File(MultipleSequentialTestExecutor::class.java.getResource("package.json").path)
```
Comment (CryDeb) [inline src/main/kotlin/ApplicationStarter.kt]: Allenfals neue Util klasse?

File: src/main/kotlin/MultipleSequentialTestExecutor.kt [hunk 2]
```diff
@@ -1 +1,51 @@
...
+        while (true) {
+            displayHeader()
+            guiLess.display("---- Tests ----")
+            testDataList.iterator().forEach { testData -> guiLess.display("${testData.id} - ${testData.name}") }
+            guiLess.getUserInput("Choose test (nr), or quit (q)")
+        }
+    }
+
+    fun quit() {
+        // TODO Implement report generation
```
Comment (CryDeb) [inline src/main/kotlin/MultipleSequentialTestExecutor.kt]: del TODOs

File: src/main/kotlin/MultipleSequentialTestExecutor.kt [hunk 3]
```diff
@@ -1 +1,51 @@
...
+    }
+
+    fun executeTest(testId: Int) {
+        guiLess.display("Execute test $testId")
+    }
+
+    override fun handleUserInput(userMessage: String) {
+        try {
+            val testId: Int = userMessage.toInt()
+            testDataList.iterator().forEach { testData ->
```
Comment (CryDeb) [inline src/main/kotlin/MultipleSequentialTestExecutor.kt]: if(!testDataList.stream().filter(data -> data.id == testId).empty()){
    executeTest(testId)
}

File: src/main/kotlin/gui/GuiLess.kt [hunk 4]
```diff
@@ -1,3 +1,28 @@
 package gui
 
-class GuiLess : Gui
\ No newline at end of file
+class GuiLess(private val commandLineReader: ScannerAbstraction) : Gui {
+    private val observers: ArrayList<GuiObserver> = ArrayList()
+
+    override fun displayMultiple(messages: List<String>) {
+        val messageIterator = messages.iterator()
+        messageIterator.forEach { message -> println(message) }
```
Comment (CryDeb) [inline src/main/kotlin/gui/GuiLess.kt]: messageIterator.forEach(System.out::println)

File: src/test/kotlin/gui/GuiLessTest.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    fun display() {
+        val guiLess = GuiLess(scannerAbstractionImpl)
+        val message = "This is a test"
+        guiLess.display(message)
+        assert(Pattern.compile(message).matcher(resultStream.toString()).find())
+    }
+
+    @Test
+    fun testSubscriptionAndFunctionality() {
+        val guiObserver = GuiObserverStub()
```
Comment (CryDeb) [inline src/test/kotlin/gui/GuiLessTest.kt:51]: // arrange

File: src/test/kotlin/gui/GuiLessTest.kt [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        val question = "Option 1?"
+        val answer = "Yes"
+        val scannerStub = ScannerStub(answer)
+        val guiLess = GuiLess(scannerStub)
+        guiLess.subscribe(guiObserver)
+        guiLess.getUserInput(question)
```
Comment (CryDeb) [inline src/test/kotlin/gui/GuiLessTest.kt:58]: // act

File: src/test/kotlin/gui/GuiLessTest.kt [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        assert(Pattern.compile(question).matcher(resultStream.toString()).find())
+        assertEquals(answer,guiObserver.userMessages.last())
+        System.setIn(originalIn)
+        guiLess.unsubscribe(guiObserver)
+        scannerStub.returnValueForNext = "No"
+        guiLess.getUserInput(question)
+        assertEquals(answer, guiObserver.userMessages.last())
```
Comment (CryDeb) [inline src/test/kotlin/gui/GuiLessTest.kt]: eigener Test

File: src/test/kotlin/gui/GuiLessTest.kt [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    @Test
+    fun testSubscriptionAndFunctionality() {
+        val guiObserver = GuiObserverStub()
+        val question = "Option 1?"
+        val answer = "Yes"
+        val scannerStub = ScannerStub(answer)
+        val guiLess = GuiLess(scannerStub)
+        guiLess.subscribe(guiObserver)
+        guiLess.getUserInput(question)
+        assert(Pattern.compile(question).matcher(resultStream.toString()).find())
```
Comment (CryDeb) [inline src/test/kotlin/gui/GuiLessTest.kt]: CleanCode?
Züge eines Integrationstest

File: src/test/kotlin/gui/GuiLessTest.kt (hunk 5 above)
Comment (CryDeb) [reply src/test/kotlin/gui/GuiLessTest.kt:51]: assertThat(resultStream.toString(), contains(message))

**Review by CryDeb — CHANGES_REQUESTED** [review]

**Comment by KilnOfTheSecondFlame** [comment]
Thanks for all of the valuable input!
I tried to implement all of it, could you please recheck and give feedback?

**Review by CryDeb — APPROVED** [review]

**Pull request merged**